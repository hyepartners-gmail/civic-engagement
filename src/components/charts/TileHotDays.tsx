"use client";

import { useMemo } from 'react';
import { BarChart, Bar, XAxis, YAxis, Tooltip, ResponsiveContainer } from 'recharts';

interface TileHotDaysProps {
  series: [number, number | null][];
  syncedDomain?: [number, number] | null;
  threshold: number;
}

export default function TileHotDays({ series, syncedDomain, threshold }: TileHotDaysProps) {
  // Convert series to Recharts format
  const chartData = useMemo(() => {
    return series
      .filter(([_, value]) => value !== null)
      .map(([year, value]) => ({
        year,
        days: value
      }));
  }, [series]);
  
  // Determine domain
  const domain = useMemo(() => {
    if (syncedDomain) return syncedDomain;
    
    if (chartData.length === 0) return [0, 10];
    
    const values = chartData.map(d => d.days as number);
    const max = Math.max(...values);
    
    return [0, max + max * 0.1];
  }, [chartData, syncedDomain]);
  
  if (chartData.length === 0) {
    return (
      <div className="h-32 flex items-center justify-center text-gray-500">
        <p>No hot days data available</p>
      </div>
    );
  }

  return (
    <div>
      <h4 className="text-sm font-medium text-gray-700 mb-2">{`Hot Days (≥${threshold}°F)`}</h4>
      <div className="h-32">
        <ResponsiveContainer width="100%" height="100%">
          <BarChart data={chartData}>
            <XAxis 
              dataKey="year" 
              hide={true}
            />
            <YAxis 
              domain={domain}
              hide={true}
            />
            <Tooltip 
              formatter={(value) => [`${value} days`, `≥${threshold}°F`]}
              labelFormatter={(year) => `Year: ${year}`}
            />
            <Bar 
              dataKey="days" 
              fill="#ef4444" 
            />
          </BarChart>
        </ResponsiveContainer>
      </div>
      <div className="flex justify-between text-xs text-gray-500 mt-1">
        <span>{chartData[0]?.year}</span>
        <span>{chartData[chartData.length - 1]?.year}</span>
      </div>
    </div>
  );
}