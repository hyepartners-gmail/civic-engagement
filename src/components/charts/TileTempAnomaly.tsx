"use client";

import { useMemo } from 'react';
import { LineChart, Line, XAxis, YAxis, Tooltip, ResponsiveContainer } from 'recharts';

interface TileTempAnomalyProps {
  series: [number, number | null][];
  syncedDomain?: [number, number] | null;
}

export default function TileTempAnomaly({ series, syncedDomain }: TileTempAnomalyProps) {
  // Convert series to Recharts format
  const chartData = useMemo(() => {
    return series
      .filter(([_, value]) => value !== null)
      .map(([year, value]) => ({
        year,
        anomaly: value
      }));
  }, [series]);
  
  // Determine domain
  const domain = useMemo(() => {
    if (syncedDomain) return syncedDomain;
    
    if (chartData.length === 0) return [0, 1];
    
    const values = chartData.map(d => d.anomaly as number);
    const min = Math.min(...values);
    const max = Math.max(...values);
    const range = max - min;
    
    // Ensure domain includes 0 for proper baseline
    return [Math.min(min - range * 0.1, 0), Math.max(max + range * 0.1, 0)];
  }, [chartData, syncedDomain]);
  
  if (chartData.length === 0) {
    return (
      <div className="h-32 flex items-center justify-center text-gray-500">
        <p>No temperature anomaly data available</p>
      </div>
    );
  }

  return (
    <div>
      <h4 className="text-sm font-medium text-gray-700 mb-2">Temperature Anomaly</h4>
      <div className="h-32">
        <ResponsiveContainer width="100%" height="100%">
          <LineChart data={chartData}>
            <XAxis 
              dataKey="year" 
              hide={true}
            />
            <YAxis 
              domain={domain}
              hide={true}
            />
            <Tooltip 
              formatter={(value) => [`${Number(value).toFixed(2)}Â°C`, 'Anomaly']}
              labelFormatter={(year) => `Year: ${year}`}
            />
            <Line 
              type="monotone" 
              dataKey="anomaly" 
              stroke="#3b82f6" 
              strokeWidth={2}
              dot={false}
              activeDot={{ r: 4 }}
            />
          </LineChart>
        </ResponsiveContainer>
      </div>
      <div className="flex justify-between text-xs text-gray-500 mt-1">
        <span>{chartData[0]?.year}</span>
        <span>{chartData[chartData.length - 1]?.year}</span>
      </div>
    </div>
  );
}