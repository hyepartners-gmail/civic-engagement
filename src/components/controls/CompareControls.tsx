"use client";

import { useState } from 'react';
import { Button } from '@/components/ui/button';
import { Switch } from '@/components/ui/switch';
import { Label } from '@/components/ui/label';
import { 
  Select, 
  SelectContent, 
  SelectItem, 
  SelectTrigger, 
  SelectValue 
} from '@/components/ui/select';
import { X } from 'lucide-react';

interface CompareControlsProps {
  cities: string[] | string;
  onCitiesChange: (cities: string[]) => void;
  syncY: boolean;
  onSyncYChange: (syncY: boolean) => void;
  threshold: number;
  onThresholdChange: (threshold: number) => void;
  allCities: Record<string, string>;
}

export default function CompareControls({
  cities,
  onCitiesChange,
  syncY,
  onSyncYChange,
  threshold,
  onThresholdChange,
  allCities
}: CompareControlsProps) {
  // Ensure cities is always an array
  const citiesArray = Array.isArray(cities) ? cities : (cities ? [cities] : []);
  
  const [newCity, setNewCity] = useState('');
  
  const addCity = () => {
    if (newCity && !citiesArray.includes(newCity)) {
      onCitiesChange([...citiesArray, newCity]);
      setNewCity('');
    }
  };
  
  const removeCity = (cityId: string) => {
    onCitiesChange(citiesArray.filter(id => id !== cityId));
  };
  
  const availableCities = Object.keys(allCities).filter(id => !citiesArray.includes(id));

  return (
    <div className="space-y-6">
      <div>
        <Label className="text-sm font-medium mb-2 block">Selected Cities</Label>
        <div className="flex flex-wrap gap-2 mb-2">
          {citiesArray.map(cityId => (
            <div 
              key={cityId} 
              className="flex items-center bg-platform-contrast/50 border border-platform-accent/30 rounded-full pl-3 pr-2 py-1 text-sm"
            >
              <span className="text-[#12001a] font-medium">{allCities[cityId] || cityId}</span>
              <button 
                onClick={() => removeCity(cityId)}
                className="ml-2 text-platform-accent hover:text-platform-accent/80"
                aria-label={`Remove ${allCities[cityId] || cityId}`}
              >
                <X size={16} />
              </button>
            </div>
          ))}
        </div>
        
        {availableCities.length > 0 && (
          <div className="flex items-center gap-2">
            <Select value={newCity} onValueChange={setNewCity}>
              <SelectTrigger className="w-48">
                <SelectValue placeholder="Add a city" />
              </SelectTrigger>
              <SelectContent>
                {availableCities.map(cityId => (
                  <SelectItem key={cityId} value={cityId}>
                    {allCities[cityId] || cityId}
                  </SelectItem>
                ))}
              </SelectContent>
            </Select>
            <Button onClick={addCity} disabled={!newCity}>Add</Button>
          </div>
        )}
      </div>
      
      <div className="flex flex-wrap items-center gap-6">
        <div className="flex items-center space-x-2">
          <Label htmlFor="sync-y">Sync Y-Axis</Label>
          <Switch
            id="sync-y"
            checked={syncY}
            onCheckedChange={onSyncYChange}
          />
        </div>
        
        <div className="flex items-center space-x-2">
          <Label htmlFor="threshold">Hot Days Threshold</Label>
          <Select 
            value={String(threshold)} 
            onValueChange={(value) => onThresholdChange(Number(value))}
          >
            <SelectTrigger className="w-24">
              <SelectValue />
            </SelectTrigger>
            <SelectContent>
              <SelectItem value="90">90°F</SelectItem>
              <SelectItem value="95">95°F</SelectItem>
              <SelectItem value="100">100°F</SelectItem>
            </SelectContent>
          </Select>
        </div>
      </div>
    </div>
  );
}
