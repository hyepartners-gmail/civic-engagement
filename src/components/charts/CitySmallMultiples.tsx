"use client";

import { useMemo } from 'react';
import TileTempAnomaly from './TileTempAnomaly';
import TileHotDays from './TileHotDays';
import { Card } from '@/components/shared/Card';

interface CitySmallMultiplesProps {
  cityData: Record<string, {
    tempAnomaly: [number, number | null][];
    hotDays: [number, number | null][];
  }>;
  cityNames: Record<string, string>;
  syncY: boolean;
  yearRange: [number, number] | null;
  threshold: number;
}

// Add interface for synced domains
interface SyncedDomains {
  tempAnomaly: [number, number] | null;
  hotDays: [number, number] | null;
}

export default function CitySmallMultiples({ 
  cityData, 
  cityNames,
  syncY,
  yearRange,
  threshold
}: CitySmallMultiplesProps) {
  // Compute synced domains if syncY is enabled
  const syncedDomains = useMemo<SyncedDomains | null>(() => {
    if (!syncY) return null;
    
    const tempAnomalySeries: Record<string, [number, number | null][]> = {};
    const hotDaysSeries: Record<string, [number, number | null][]> = {};
    
    Object.entries(cityData).forEach(([cityId, data]) => {
      tempAnomalySeries[cityId] = data.tempAnomaly;
      hotDaysSeries[cityId] = data.hotDays;
    });
    
    return {
      tempAnomaly: computeSyncedDomain(tempAnomalySeries),
      hotDays: computeSyncedDomain(hotDaysSeries)
    };
  }, [cityData, syncY]);
  
  // Helper function to compute domain for a set of series
  const computeSyncedDomain = (seriesByCity: Record<string, [number, number | null][]>) => {
    let globalMin = Infinity;
    let globalMax = -Infinity;
    let hasValidData = false;
    
    Object.values(seriesByCity).forEach(series => {
      series.forEach(([_, value]) => {
        if (value !== null) {
          hasValidData = true;
          globalMin = Math.min(globalMin, value);
          globalMax = Math.max(globalMax, value);
        }
      });
    });
    
    // If no valid data, return a default domain
    if (!hasValidData) {
      return [-1, 1] as [number, number];
    }
    
    // Add some padding
    const range = globalMax - globalMin;
    const padding = range > 0 ? range * 0.1 : 1;
    
    // Handle edge case where all values are the same
    if (range === 0) {
      return [globalMin - 1, globalMax + 1] as [number, number];
    }
    
    return [globalMin - padding, globalMax + padding] as [number, number];
  };
  
  // Trim data based on year range
  const trimmedCityData = useMemo(() => {
    if (!yearRange) return cityData;
    
    const [startYear, endYear] = yearRange;
    const result: Record<string, {
      tempAnomaly: [number, number | null][];
      hotDays: [number, number | null][];
    }> = {};
    
    Object.entries(cityData).forEach(([cityId, data]) => {
      result[cityId] = {
        tempAnomaly: data.tempAnomaly.filter(([year]) => year >= startYear && year <= endYear),
        hotDays: data.hotDays.filter(([year]) => year >= startYear && year <= endYear)
      };
    });
    
    return result;
  }, [cityData, yearRange]);

  return (
    <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-6">
      {Object.entries(trimmedCityData).map(([cityId, data]) => (
        <Card key={cityId} className="p-4">
          <h3 className="text-lg font-semibold mb-3 text-center">{cityNames[cityId] || cityId}</h3>
          <div className="space-y-4">
            <TileTempAnomaly 
              series={data.tempAnomaly} 
              syncedDomain={syncY ? syncedDomains?.tempAnomaly : undefined}
            />
            <TileHotDays 
              series={data.hotDays} 
              syncedDomain={syncY ? syncedDomains?.hotDays : undefined}
              threshold={threshold}
            />
          </div>
        </Card>
      ))}
    </div>
  );
}