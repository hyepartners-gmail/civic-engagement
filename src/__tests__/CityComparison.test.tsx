import { render } from '@testing-library/react';
import CitySmallMultiples from '@/components/charts/CitySmallMultiples';
import TileTempAnomaly from '@/components/charts/TileTempAnomaly';
import TileHotDays from '@/components/charts/TileHotDays';
import CompareControls from '@/components/controls/CompareControls';

// Mock data for testing
const mockCityData = {
  seattle: {
    tempAnomaly: [
      [2000, 0.5],
      [2001, 0.6],
      [2002, 0.7]
    ],
    hotDays: [
      [2000, 5],
      [2001, 7],
      [2002, 6]
    ]
  },
  'los-angeles': {
    tempAnomaly: [
      [2000, 1.2],
      [2001, 1.3],
      [2002, 1.4]
    ],
    hotDays: [
      [2000, 25],
      [2001, 30],
      [2002, 28]
    ]
  }
};

const mockCityNames = {
  seattle: 'Seattle',
  'los-angeles': 'Los Angeles'
};

const mockAllCities = {
  seattle: 'Seattle',
  'los-angeles': 'Los Angeles',
  chicago: 'Chicago',
  houston: 'Houston'
};

describe('City Comparison Components', () => {
  describe('CitySmallMultiples', () => {
    it('should render correctly with syncY enabled', () => {
      const { container } = render(
        <CitySmallMultiples
          cityData={mockCityData}
          cityNames={mockCityNames}
          syncY={true}
          yearRange={null}
          threshold={90}
        />
      );
      
      expect(container).toMatchSnapshot();
    });

    it('should render correctly with syncY disabled', () => {
      const { container } = render(
        <CitySmallMultiples
          cityData={mockCityData}
          cityNames={mockCityNames}
          syncY={false}
          yearRange={null}
          threshold={90}
        />
      );
      
      expect(container).toMatchSnapshot();
    });
  });

  describe('TileTempAnomaly', () => {
    it('should render correctly with data', () => {
      const { container } = render(
        <TileTempAnomaly 
          series={mockCityData.seattle.tempAnomaly} 
          syncedDomain={null}
        />
      );
      
      expect(container).toMatchSnapshot();
    });

    it('should render correctly with synced domain', () => {
      const { container } = render(
        <TileTempAnomaly 
          series={mockCityData.seattle.tempAnomaly} 
          syncedDomain={[-1, 2]}
        />
      );
      
      expect(container).toMatchSnapshot();
    });
  });

  describe('TileHotDays', () => {
    it('should render correctly with data', () => {
      const { container } = render(
        <TileHotDays 
          series={mockCityData.seattle.hotDays} 
          syncedDomain={null}
          threshold={90}
        />
      );
      
      expect(container).toMatchSnapshot();
    });

    it('should render correctly with synced domain', () => {
      const { container } = render(
        <TileHotDays 
          series={mockCityData.seattle.hotDays} 
          syncedDomain={[0, 50]}
          threshold={90}
        />
      );
      
      expect(container).toMatchSnapshot();
    });
  });

  describe('CompareControls', () => {
    it('should render correctly', () => {
      const { container } = render(
        <CompareControls
          cities={['seattle', 'los-angeles']}
          onCitiesChange={jest.fn()}
          syncY={true}
          onSyncYChange={jest.fn()}
          threshold={90}
          onThresholdChange={jest.fn()}
          allCities={mockAllCities}
        />
      );
      
      expect(container).toMatchSnapshot();
    });
  });
});