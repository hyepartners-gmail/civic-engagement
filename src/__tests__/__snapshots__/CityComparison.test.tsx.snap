// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`City Comparison Components CompareControls should render correctly 1`] = `
<div>
  <div
    class="space-y-6"
  >
    <div>
      <label
        class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-sm font-medium mb-2 block"
      >
        Selected Cities
      </label>
      <div
        class="flex flex-wrap gap-2 mb-2"
      >
        <div
          class="flex items-center bg-platform-contrast/50 border border-platform-accent/30 rounded-full pl-3 pr-2 py-1 text-sm"
        >
          <span
            class="text-[#12001a] font-medium"
          >
            Seattle
          </span>
          <button
            aria-label="Remove Seattle"
            class="ml-2 text-platform-accent hover:text-platform-accent/80"
          >
            <svg
              class="lucide lucide-x"
              fill="none"
              height="16"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18 6 6 18"
              />
              <path
                d="m6 6 12 12"
              />
            </svg>
          </button>
        </div>
        <div
          class="flex items-center bg-platform-contrast/50 border border-platform-accent/30 rounded-full pl-3 pr-2 py-1 text-sm"
        >
          <span
            class="text-[#12001a] font-medium"
          >
            Los Angeles
          </span>
          <button
            aria-label="Remove Los Angeles"
            class="ml-2 text-platform-accent hover:text-platform-accent/80"
          >
            <svg
              class="lucide lucide-x"
              fill="none"
              height="16"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18 6 6 18"
              />
              <path
                d="m6 6 12 12"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="flex items-center gap-2"
      >
        <button
          aria-autocomplete="none"
          aria-controls="radix-:r0:"
          aria-expanded="false"
          class="flex h-10 items-center justify-between rounded-md border border-platform-accent bg-platform-contrast px-3 py-2 text-platform-text text-sm focus:outline-none focus:ring-2 focus:ring-platform-accent focus:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 w-48"
          data-placeholder=""
          data-state="closed"
          dir="ltr"
          role="combobox"
          type="button"
        >
          <span
            style="pointer-events: none;"
          >
            Add a city
          </span>
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down h-4 w-4 text-platform-accent"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
          disabled=""
          tabindex="0"
        >
          Add
        </button>
      </div>
    </div>
    <div
      class="flex flex-wrap items-center gap-6"
    >
      <div
        class="flex items-center space-x-2"
      >
        <label
          class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
          for="threshold"
        >
          Hot Days Threshold
        </label>
        <button
          aria-autocomplete="none"
          aria-controls="radix-:r1:"
          aria-expanded="false"
          class="flex h-10 items-center justify-between rounded-md border border-platform-accent bg-platform-contrast px-3 py-2 text-platform-text text-sm focus:outline-none focus:ring-2 focus:ring-platform-accent focus:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 w-24"
          data-state="closed"
          dir="ltr"
          role="combobox"
          type="button"
        >
          <span
            style="pointer-events: none;"
          >
            90Â°F
          </span>
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down h-4 w-4 text-platform-accent"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
